client/types/callbacks.go

```diff
@@ -110,6 +111,19 @@ type Callbacks interface {
 
 	// OnCommand is called when the Server requests that the connected Agent perform a command.
 	OnCommand(ctx context.Context, command *protobufs.ServerToAgentCommand) error
+
+	// CheckRedirect is called before following a redirect, allowing the client
+	// the opportunity to observe the redirect chain, and optionally terminate
+	// following redirects early.
+	//
+	// CheckRedirect is intended to be similar, although not exactly equivalent,
+	// to net/http.Client's CheckRedirect feature. Unlike in net/http, the via
+	// parameter is a slice of HTTP responses, instead of requests. This gives
+	// an opportunity to users to know what the exact response headers and
+	// status were. The request itself can be obtained from the response.
+	//
+	// The responses in the via parameter are passed with their bodies closed.
+	CheckRedirect(req *http.Request, via []*http.Response) error
```

Comment:
Is it valid to return `ErrUseLastResponse`? I assume it is not valid at least for WS transport since it typically means there is no WS connection established.

Reply:
For clarity we should probably also link to https://pkg.go.dev/net/http#Client.CheckRedirect somewhere in the comment above.

Reply:
This add `CheckRedirect` to `Callbacks` but implements it only for `wsClient`. I think we also need an implementation for `httpClient`.

Reply:
@tigrannajaryan Thank you for the review! 

Regarding `ErrUseLastResponse`, I would agree it is not applicable here. I think it should result in the same behaviour as returning any other error, however. I'll mention it specifically in the docs.

I agree that we should have an implementation for `httpClient`. Perhaps the name of this callback should be `WSCheckRedirect` or similar, denoting that it is only for websocket clients. I don't think the function signature, as implemented, is valid for HTTP clients.

I looked at `httpsender` and it uses the default HTTP client (with an override method for TLS config). Perhaps `httpClient` could accept a `net/http.Client` supplied by the library consumer, instead of having a specific API for things like CheckRedirect. Then the API surface would not grow, and library consumers would be able to use any custom configuration of the HTTP client they desire. Any thoughts on this?

Reply:
HTTP clients should rightly expect to have the signature `func(*http.Request, []*http.Request) error`, as that would be 1:1 with `net/http`.

Reply:
> HTTP clients should rightly expect to have the signature `func(*http.Request, []*http.Request) error`, as that would be 1:1 with `net/http`.

Do we have the need to use a different signature for WS? I see it is mentioned in the comment but it is not clear to me how exactly you would use the response for WS and why it is useful for WS and not for http.

Ideally we should use the same signature as the standard lib and make this setting applicable to both WS and plain HTTP implementations.

Reply:
I think there is a need for situations where the opamp-go users wants to know what status the server wrote. 301 vs 302 for instance, where the client may want to cache permanent redirects but not cache temporary redirects. Since the opamp-go user isn't privy to the HTTP response, there is no other opportunity to know that information.

I'm definitely open to another design, as I agree it would be better to maintain parity with net/http in general.

Reply:
@echlebek  Sorry for long pause, I finally got back to this PR.

What do you think if we change `CheckRedirect` to look like this:
```go
CheckRedirect(req *http.Request, viaReq []*http.Request, viaResp []*http.Response) error
```

The wsClient will pass both viaReq and viaResp parameters, while httpClient will only pass viaReq and will leave viaResp nil.

Reply:
Sorry for the even longer pause! That suggestion sounds good to me, I will rebase this PR and make the change.